### DISCUSSION TITLE
[Authentication] implement SAML authentication mechanism

### DISCUSSION THREAD

**Discussion by pzsette** [mia-platform/community#170] [Q&A] [answered]
Hi all,
For a client we have the need to implement SPID as authentication mechanism. Since SPID is based on SAML2 we’re trying to integrate it with the current MIA services based on OAuth2.
We thought about two different options:
- Delegate authentication to SAML and authorization to OAuth. The Client gets a SAML bearer assertion from the SAML Identity Provider then requests an access token from the Authorization Server using the SAML bearer assertion as proof of identity. In this we will probably need to evolve the `Auth0 client` to let him work with the identity provider.
https://wso2.com/blogs/cloud/integrating-oauth-api-gateway-with-saml-identity-provider/
- Build a specific service to allow both authentication and authorization with SAML. 
In this case probably the `authorization-service` should be able to call the right service to retrieve the information about the current logged user. 

Do you have any suggestions on what could be the best strategy to implement?
Has someone already used an alternative identity provider integrated in the product authN and authZ flow?

**Comment by davidebianchi** [comment, marked as answer]
Hi @pzsette, the question is not so related with Mia-Platform, it is more an architectural decision which depends from all the context so it is quite difficult to answer.

In the first case, the auth0-client as stated by the name enable integration only with Auth0, so the direct integration with SPID is not an option from my point of view. So, maybe it is possible to enable the integration using Auth0? [Here](https://auth0.com/docs/authenticate/protocols/saml/saml-configuration/configure-auth0-as-service-and-identity-provider) a documentation page, but I don't know how much it is related. I don't really know also if there are some specific security or license constraints which may not follow this path.

In the second case, it is possible if you use JWT and envoy as api-gateway and differentiate based on JWT claims which authorization-service to use. Otherwise, if you use nginx with the authorization-service, you should set as `USERINFO_URL` of the authorization-service a custom service, which discriminate based on the authorization header which idp to contact to resolve the user which make the call.

I hope that this could help you to take the decision.
